File: MemoryProxy.cpp
//
// Created by alexs on 4/13/2025.
//

#include "MemoryProxy.h"

MemoryProxy::MemoryProxy(RAM *memory): ram(memory), pageTable(nullptr) {}

Word& MemoryProxy::readWord(const int& address) {
    int block = address / BLOCK_SIZE;
    int index = address % BLOCK_SIZE;
    int realBlock = pageTable->data[block].toInteger();
    return ram->readWord(realBlock, index);
}

void MemoryProxy::writeWord(Word word, const int& address) {
    int block = address / BLOCK_SIZE;
    int index = address % BLOCK_SIZE;
    int realBlock = pageTable->data[block].toInteger();
    ram->writeWord(word, realBlock, index);
}

void MemoryProxy::setPageTable(MemoryBlock *pageTable) {
    this->pageTable = pageTable;
}

void MemoryProxy::printBlock(int block) {
    int realBlock = pageTable->data[block].toInteger();
    ram->printBlock(realBlock);
}

File: src/MemoryProxy.java
public class MemoryProxy {
    private final RAM ram;
    private MemoryBlock pageTable;

    public MemoryProxy(RAM ram) {
        this.ram = ram;
    }

    public void setPageTable(MemoryBlock pageTable) {
        this.pageTable = pageTable;
    }

    public void writeWord(Word word, int address) {
        int block = address / Constants.BLOCK_SIZE;
        int index = address % Constants.BLOCK_SIZE;
        int realBlock = pageTable.data[block].toInteger();
        ram.writeWord(word, realBlock, index);
    }

    public Word readWord(int address) {
        int block = address / Constants.BLOCK_SIZE;
        int index = address % Constants.BLOCK_SIZE;
        int realBlock = pageTable.data[block].toInteger();
        return ram.readWord(realBlock, index);
    }

    public void printBlock(int block) {
        int realBlock = pageTable.data[block].toInteger();
        ram.printBlock(realBlock);
    }
}
